File: src/main/java/com/tsunazumi/algorithms/hackerrank/algorithms/archive/CutTheSticks.java
package com.tsunazumi.algorithms.hackerrank.algorithms.archive;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CutTheSticks {
  public static void main(String[] args) {
    List<Integer> input = Arrays.asList(5,4,4,2,2,8);
    List<Integer> result = new ArrayList<>();
    List<Integer> finalResult = cutTheSticks(input, result);
    for (int i : finalResult) {
      System.out.println(i);
    }
  }

  public static List<Integer> cutTheSticks(List<Integer> arr, List<Integer> result) {
    if (arr.size() == 0) {
      return result;
    }

    // find the smallest number
    int smallestNumber = arr.get(0);
    for (int i : arr) {
      if (i < smallestNumber) {
        smallestNumber = i;
      }
    }

    // cut that amount from each item in the list
    List<Integer> newArr = new ArrayList<>();
    for (int i : arr) {
      if (i - smallestNumber > 0) {
        newArr.add(i - smallestNumber);
      }
    }
    result.add(arr.size());
    return cutTheSticks(newArr, result);
  }

}

File: src/main/java/com/tsunazumi/dsa/hackerrank/algorithms/archive/CutTheSticks.kt
package com.tsunazumi.dsa.hackerrank

fun main(arg: Array<String>) {
  val arr = arrayOf(5,4,4,2,2,8)
  var result = ArrayList<Int>()
  val finalresult = cutTheSticks(arr, result)
  finalresult.forEach { println(it) }
}

fun cutTheSticks(arr: Array<Int>, result: ArrayList<Int>): Array<Int> {
  if (arr.isEmpty()) {
    return result.toTypedArray();
  }
  var input = arr.toList()
  // find the smallest number
  val smallest = input.sorted()[0]

  // cut that amount from each item in the list
  var newArr = ArrayList<Int>()
  for (i in arr) {
    if ( i - smallest > 0) {
      newArr.add(i - smallest)
    }
  }
  result.add(arr.size)

  return cutTheSticks(newArr.toTypedArray(), result)

}